src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/CardinalityBasedGreedyJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,350 @@
+                    final List<FederationMember> fms = new ArrayList<>();
+                    for ( int count = 0; count < subplan.numberOfSubPlans(); count++ ) {
+                        cardinality += resps[index].getCardinality();
+
+                        final FederationMember fm = flattenRequestMemPairs.get(index).getMember();
+                        numOfAccess += accessNumForReq( resps[index].getCardinality(), fm );
+                        fms.add( fm );
+
+                        index++;
+                    }
+
+                    planWithStatistics = new PhysicalPlanWithStatistics(subplan, fms, cardinality, numOfAccess );
+                }
+                else
+                    throw new IllegalArgumentException("Unsupported type of subquery in source assignment (" + pop.getClass().getName() + ")");
+
+                subPlansWithStatistics.add( planWithStatistics );
+            }
+
+            return subPlansWithStatistics;
+        }
+
+        /**
+         * Compares all available subplans (see {@link #subplans}) in terms of
+         * their respective cardinality returns the one with the lowest cardinality.
+         */
+        protected PhysicalPlanWithStatistics chooseFirstSubPlan( final List<PhysicalPlanWithStatistics> subPlansWithStatistics ) {
+            PhysicalPlanWithStatistics bestPlanWithCandidate = subPlansWithStatistics.get(0);
+            int lowestCost = bestPlanWithCandidate.getCardinality();
+
+            for ( int i = 1; i < subPlansWithStatistics.size(); i++ ) {
+                PhysicalPlanWithStatistics current = subPlansWithStatistics.get(i);
+                if ( current.getCardinality() < lowestCost ) {
+                    lowestCost = current.getCardinality();
+                    bestPlanWithCandidate = current;
+                }
+            }
+
+            return bestPlanWithCandidate;
+        }
+
+        /**
+         * Iterate through the remaining subplans and selected those that contain any of the variables in the given set of variables.
+         */
+        protected List<PhysicalPlanWithStatistics> getSubPlansContainVars( final ExpectedVariables vars, final List<PhysicalPlanWithStatistics> subPlansWithStatistics ) {
+            final List<PhysicalPlanWithStatistics> subPlansContainsVars = new ArrayList<>();
+            for ( final PhysicalPlanWithStatistics subplan: subPlansWithStatistics ) {
+                if ( ! ExpectedVariablesUtils.intersectionOfAllVariables( vars, subplan.plan.getExpectedVariables() ).isEmpty() ){
+                    subPlansContainsVars.add(subplan);
+                }
+            }
+
+            if ( !subPlansWithStatistics.isEmpty() && subPlansContainsVars.isEmpty() ){
+                // Independent subplans exist. In this case, choose the subplan with the lowest candidate (from remaining subPlans) as candidate plan
+                return Arrays.asList( chooseFirstSubPlan(subPlansWithStatistics) );
```

Comment:
No, this doesn't belong into this function. Move it into the loop from where you call this function.